llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp

```diff
@@ -1323,25 +1324,43 @@ SystemZTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
 }
 
 bool SystemZTTIImpl::shouldExpandReduction(const IntrinsicInst *II) const {
-  // Always expand on Subtargets without vector instructions
+  // Always expand on Subtargets without vector instructions.
   if (!ST->hasVector())
     return true;
 
-  // Always expand for operands that do not fill one vector reg
-  auto *Type = cast<FixedVectorType>(II->getOperand(0)->getType());
-  unsigned NumElts = Type->getNumElements();
-  unsigned ScalarSize = Type->getScalarSizeInBits();
+  // Find the type of the vector operand of the intrinsic
+  // This assumes that each vector reduction intrinsic only
+  // has one vector operand.
+  FixedVectorType *VType = 0x0;
+  for (unsigned I = 0; I < II->getNumOperands(); ++I) {
+    auto *T = II->getOperand(I)->getType();
+    if (T->isVectorTy()) {
+      VType = cast<FixedVectorType>(T);
+      break;
+    }
+  }
+
```

Comment:
It doesn't seem quite foolproof to me to make the assumption that all vector operands have the same type. Or is it?

If so, one option could be to keep looping and if VType is non-null (following operands) assert that VType == T.

Reply:
I think we should either explicitly handle each intrinsic by opcode (opc -> vector type), or else at least make sure that there is only one vector operand type. If something different emerges in the future, I guess then we would have to handle it by opcode.

Reply:
I'm just thinking that, whether or not the operand vector is full is probably going to be a factor in many of the reductions. So it seems wrong to have it just in the branch for `vector.reduce.add`. I do see your point though about keeping things as simple as possible. I'm going to refactor things a bit to make things simpler, but kepp some of the extensibility in case we need to handle more intrinsics.

Reply:
shouldExpandReduction() looks nice now.

I don't quite like getFirstOperandVectorType(), it's not robust against any future instrinsics we do not know about, and it doesn't have an assert that would discover any such future cases if they emerge.

We want to know the intrinsic type, and we know which intrinsic we are dealing with. So why make it unnecessarily general? No need to loop as we know what operand number we need to use. This could be done directly in shouldExpandReduction(), or if you want to have a function for it, you could pass the operand number to it.